File: Leetcode/141.环形链表.cpp
//快慢指针实现
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        ListNode *fast = head, *slow = head;

        do {
            if (fast != nullptr && fast -> next != nullptr) {
                fast = fast -> next -> next;
                slow = slow -> next;
            } else {
                return false;
            }
        } while(fast != slow);

        return true;
    }
};

//hash表实现
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        unordered_set<ListNode*> set_;

        while (head) {
            if (set_.count(head)) {
                return true;
            }
            set_.insert(head);
            head = head -> next;
        }

        return false;
    }
};

File: Leetcode/141.环形链表.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func hasCycle(head *ListNode) bool {
	slow := head
	fast := head
	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
		if fast == slow {
			return true
		}
	}

	return false
}

func hasCycle(head *ListNode) bool {
	map_ := make(map[*ListNode]int)

	for head != nil {
		if _, ok := map_[head]; ok {
			return true
		}
		map_[head] = 1
		head = head.Next
	}

	return false
}